bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+        """
+        return self
+
+    def __next__(self) -> DrawAndLogP:
+        """Yields next draw from DrGhmcDiag iterator.
+
+        Yields:
+            DrawAndLogP: Tuple of (sample, log density)
+        """
+        return self.sample()
+
+    def joint_logp(self, theta: VectorType, rho: VectorType) -> float:
+        """Joint log density of sample (theta, rho) under the cannonical distribution.
+
+        Args:
+            theta: position
+            rho: momentum
+
+        Returns:
+            float: joint log density
+        """
+        adj: float = 0.5 * np.dot(rho, self._metric * rho)
+        return self._model.log_density(theta) - adj
+
+    def leapfrog(
+        self,
+        theta: VectorType,
+        rho: VectorType,
+        stepsize: float,
+        steps: int,
+    ) -> tuple[VectorType, VectorType]:
+        """Propose new sample (theta_prop, rho_prop) with leapfrog integrator.
+
+        Discretize and solve Hamilton's equations using the leapfrog integrator to
+        deterministically propose a new sample (theta_prop, rho_prop) from the current
+        sample (theta, rho).
+
+        Copy theta because numpy's += operator mutates the original array instead of
+        creating a new one.
+
+        Initialize rho_mid by going backwards half a step so that the first full-step
+        inside the loop brings rho_mid up to +1/2 steps.
+
+        Args:
+            theta: position
+            rho: momentum
+            stepsize: stepsize in each leapfrog step
+            steps: number of leapfrog steps
+
+        Returns:
+            tuple[VectorType, VectorType]: proposed sample
+        """
+        theta = np.array(theta, copy=True)
+        _, grad = self._model.log_density_gradient(theta)
+        rho_mid = rho - 0.5 * stepsize * np.multiply(self._metric, grad).squeeze()
+        for _ in range(steps):
+            rho_mid += stepsize * np.multiply(self._metric, grad).squeeze()
+            theta += stepsize * rho_mid
+            _, grad = self._model.log_density_gradient(theta)
+        rho = rho_mid + 0.5 * stepsize * np.multiply(self._metric, grad).squeeze()
+        return (theta, rho)
+
+    def _prob_of_delayed_rejection(self, hastings: float) -> float:
+        """Log probability of making another proposal upon rejection.
+
+        To reduce average cost per iteration for DRHMC, make the delayed rejections
+        probabilistic, such that a subsequent proposal is not mandatory upon rejection.
+
+        Probabilistic delayed rejection proposes a new sample with some probability
+        dependent on where we are in the distribution. If a proposal has low acceptance
+        probability and was rejected, propose another sample; if a proposal has high
+        acceptance probablity but was rejected by chance, do not propose another
+        sample. More details found in section 3.2 of Modi et al.
+
+        A heuristic proposal probability that achieve this is eqn. 31 in Modi et al:
```

Comment:
Doc shouldn't talk about possibilities.  If this is the version that's used, just say that.  You don't need to mention that we always make the first proposal---the algorithm wouldn't make sense otherwise, but if you want to be really clear, talk about probabilities of subsequent proposals.